shared/utility/motion/ForwardKinematics.h

```diff
@@ -324,61 +318,198 @@ namespace motion {
                 calculatePosition(model, sensors, ServoID::R_ELBOW);
             result.insert(leftArm.begin(), leftArm.end());
             result.insert(rightArm.begin(), rightArm.end());
-            result.insert(leftLegPositions.begin(), leftLegPositions.end());
             result.insert(rightLegPositions.begin(), rightLegPositions.end());
             result.insert(headPositions.begin(), headPositions.end());
             return result;
         }
         /*! @brief Adds up the mass vectors stored in the robot model and normalises the resulting position
             @return [x_com, y_com, z_com, total_mass] relative to the torso basis
         */
-        inline arma::vec4 calculateCentreOfMass(
+        inline Eigen::Vector4d calculateCentreOfMass(
             const message::motion::KinematicsModel& model,
-            const std::array<Eigen::Matrix<double, 4, 4, Eigen::DontAlign>, 20>& jointPositions,
-            bool includeTorso) {
-            arma::vec4 totalMassVector = arma::zeros(4);
-
-            for (size_t j = 0; j < 20; j++) {
-                const auto& joint = jointPositions[j];
-                arma::vec4 massVector;
-                for (size_t i = 0; i < 4; i++) {
-                    massVector[i] = model.massModel.masses[j][i];
-                }
-                // NUClear::log<NUClear::DEBUG>("calculateCentreOfMass - reading mass ",
-                // message::input::stringFromId(j), massVector);
-                double jointMass = massVector[3];
-
-                utility::math::matrix::Transform3D massScaler;
-                massScaler.submat(0, 0, 2, 2) *= jointMass;
-
-                // = m * local centre of mass in global robot coords
-                totalMassVector += convert<double, 4, 4>(joint) * massScaler * massVector;
+            const std::array<Eigen::Matrix<double, 4, 4, Eigen::DontAlign>, 20>& forward_kinematics,
+            bool include_torso,
+            const Eigen::Matrix4d& Hwt,
+            bool debug = false) {
+
+            // Convenience function to transform particle-space CoM to torso-space CoM
+            // Htp - transform from particle space to torso space
+            // particle - CoM coordinates in particle space
+            // invert - (TODO) in case left/right symmetry is broken
```

Comment:
Can this TODO be resolved?

Reply:
invert was to account for potential differences between the left thigh model being used for the right thigh (they arent exact mirror duplicates of each other, same for the rest of the leg and arm). However, CoM seems to be working fine, so I don't know if its worth the time to properly invert these parts at this time